### DISCUSSION TITLE
What is the mathematical principle of the sensorsAccAlignToGravity function?

### DISCUSSION THREAD

**Discussion by HelloIDTC** [bitcraze/discussions#1359] [Q&A]
What is the mathematical principle of the sensorsAccAlignToGravity function?
How do I know the roll and pitch angles of the accelerometer in its initial state?
https://github.com/bitcraze/crazyflie-firmware/blob/master/src/hal/src/sensors_bmi088_bmp3xx.c#L926

**Comment by knmcguire** [comment]
So the function name is a bit misleading. What the function actually does is to compensate the accelerometers with the roll pitch trim values that is set by the user in the cfclient. This is usually done when the Crazyflie in manual flight hovering, has an noticeable bias towards one direction. The roll  pitch trim values can be used to compensate for that.

About the mathematical principle, this seems to just rotate the accelerometers  axis with the roll pitch trim values by using rotation calculation, which seems legit. 

```
/**
 * Compensate for a miss-aligned accelerometer. It uses the trim
 * data gathered from the UI and written in the config-block to
 * rotate the accelerometer to be aligned with gravity.
 */
```

**Reply by HelloIDTC** [reply]
OK, but can I first use the difference between the accelerometer value and the gravity vector as the accelerometer tilt offset, and then subtract this tilt offset when actually using the accelerometer, like this:
` 
	float gravity[] = {0, 0, 1.0f};
	float acc[3];
	float offset[3];

	 // initialize
	readAcc(acc);
	
	offset[0] = acc[0] - gravity[0];
	offset[1] = acc[1] - gravity[1];
	offset[2] = acc[2] - gravity[2];	

	// loop
	while (1) {
		readAcc(acc);
	
		acc[0] = acc[0] - offset[0];
		acc[1] = acc[1] - offset[1];
		acc[2] = acc[2] - offset[2];
	}
`

**Reply by knmcguire** [reply]
I don't think that that will go well in flight. The accelerometers are body fixed and the gravity vectors are fixed to the world so these need to be rotated to the body fixed coordinate system as well.

If you want any inspiration, perhaps it is good to check out the kalman filter's way of handling these as well. You can find it here https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/kalman_core/kalman_core.c#L506-L509

**Comment by HelloIDTC** [comment]
But I found that this sensorsAccAlignToGravity function does not work well in the case of large angles. I tried to tilt the accelerometer by about 30 degrees and apply this function, but the results did not align with gravity.

**Reply by tobbeanton** [reply]
There is another mechanism if you would like to change the orientation of the IMU: [IMU_PHI, IMU_THETA, IMU_PSI](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/platform/interface/platform_defaults.h#L58). The sensorsAccAlignToGravity I think has only been design to do "trimming".

**Reply by HelloIDTC** [reply]
Trimming means correcting the tilt at a small angle? Should I use the sensorsAlignToAirframe function if the tilt angle is large?
https://github.com/bitcraze/crazyflie-firmware/blob/master/src/hal/src/sensors_bmi088_bmp3xx.c#L894
Is that the main difference between these two functions?
Can I use the sensorsAlignToAirframe function instead of the sensorsAccAlignToGravity function for any tilt roll, pitch, yaw angle?
But how do I know the tilt roll, pitch, yaw angle of the accelerometer in its initial state?
Can the tilt roll, pitch, yaw angle be calculated automatically and the alignment done automatically?

**Reply by tobbeanton** [reply]
>  Should I use the sensorsAlignToAirframe function if the tilt angle is large?

The sensorsAccAlignToGravity only "trims" the accelerometer while sensorsAlignToAirframe aligns the full IMU (gyro+acc). The sensorsAlignToAirframe should be used when the IMU is not mounted as on a standard Crazyflie.

> Can the tilt roll, pitch, yaw angle be calculated automatically and the alignment done automatically?

No functionality like this exist. You could try and implement it if you like.
